File: src/insert/tablet.js
import path from "path";
import csv from "papaparse";
import { mow } from "../record.js";

export async function insertTablet(fs, dir, tablet, query) {
  const filepath = path.join(dir, tablet.filename);

  // filter out the { _: trunk, trunk: value }
  // which mow returns when there's no connections
  const grains = mow(query, tablet.trunk, tablet.branch).filter(
    (grain) => grain[tablet.branch] !== undefined,
  );

  const lines = grains.map(
    ({ [tablet.trunk]: key, [tablet.branch]: value }) => {
      const keyEscaped = escape(key);

      const valueEscaped = escape(value);

      return csv.unparse([[keyEscaped, valueEscaped]], {
        delimiter: ",",
        newline: "\n",
      });
    },
  );

  for (const line of lines) {
    await fs.promises.appendFile(filepath, line + "\n");
  }
}

File: src/dataset/insert/tablet.rs
use std::path::PathBuf;
use std::fs::File;
use std::fs;
use std::fs::OpenOptions;
use crate::{line::Line, Entry, Result};
use super::Tablet;

pub async fn insert_tablet(path: PathBuf, tablet: Tablet, entry: Entry) -> Result<()> {
    let filepath = path.join(&tablet.filename);

    // create file if it doesn't exist
    if fs::metadata(&filepath).is_err() {
        File::create(&filepath)?;
    }

    let file = OpenOptions::new()
        .append(true)
        .open(filepath)
        .expect("cannot open file");

    let mut wtr = csv::WriterBuilder::new()
        .has_headers(false)
        .from_writer(file);

    let grains = entry.mow(&tablet.trunk, &tablet.branch);

    let lines: Vec<Line> = grains.iter().filter(|grain| grain.leaf_value.is_some()).map(|grain| {
        Line {
            key: match &grain.base_value { None => String::from(""), Some(s) => s.to_owned() },
            value: match &grain.leaf_value { None => String::from(""), Some(s) => s.to_owned() }
        }
    }).collect();

    for line in lines.iter() {
        wtr.serialize(line.escape())?;
    }

    Ok(())
}
